File: subly_ios/Subly/Subly/Domain/UseCases/GetCategorySpendUseCase.swift
import Foundation
import Combine

struct CategorySpend: Identifiable {
    let category: Category
    let monthlyAmount: Double
    let percentage: Float  // 0.0–1.0 share of total monthly spend
    var id: UUID { category.id }
}

final class GetCategorySpendUseCase {
    private let subscriptionRepository: SubscriptionRepository
    private let categoryRepository: CategoryRepository

    init(
        subscriptionRepository: SubscriptionRepository,
        categoryRepository: CategoryRepository
    ) {
        self.subscriptionRepository = subscriptionRepository
        self.categoryRepository = categoryRepository
    }

    func execute() -> AnyPublisher<[CategorySpend], Never> {
        Publishers.CombineLatest(
            subscriptionRepository.getActiveSubscriptions(),
            categoryRepository.getAllCategories()
        )
        .map { subscriptions, categories in
            let categoryMap = Dictionary(uniqueKeysWithValues: categories.map { ($0.id, $0) })

            let breakdown = Dictionary(
                grouping: subscriptions,
                by: { $0.categoryId }
            ).mapValues { subs in
                subs.reduce(0.0) { $0 + $1.monthlyAmount }
            }

            let total = max(breakdown.values.reduce(0, +), 1.0)

            return breakdown.compactMap { categoryId, amount -> CategorySpend? in
                guard let category = categoryMap[categoryId] else { return nil }
                return CategorySpend(
                    category: category,
                    monthlyAmount: amount,
                    percentage: Float(amount / total)
                )
            }
            .sorted { $0.monthlyAmount > $1.monthlyAmount }
        }
        .eraseToAnyPublisher()
    }
}

File: subly_android/app/src/main/java/net/cynreub/subly/domain/usecase/GetCategorySpendUseCase.kt
package net.cynreub.subly.domain.usecase

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.combine
import net.cynreub.subly.domain.model.BillingFrequency
import net.cynreub.subly.domain.model.Category
import net.cynreub.subly.domain.model.Subscription
import net.cynreub.subly.domain.repository.CategoryRepository
import net.cynreub.subly.domain.repository.SubscriptionRepository
import javax.inject.Inject

data class CategorySpend(
    val category: Category,
    val monthlyAmount: Double,
    val percentage: Float   // 0..1 share of total monthly spend
)

class GetCategorySpendUseCase @Inject constructor(
    private val subscriptionRepository: SubscriptionRepository,
    private val categoryRepository: CategoryRepository
) {
    operator fun invoke(): Flow<List<CategorySpend>> =
        combine(
            subscriptionRepository.getActiveSubscriptions(),
            categoryRepository.getAllCategories()
        ) { subscriptions, categories ->
            val categoryMap = categories.associateBy { it.id }

            val breakdown = subscriptions
                .groupBy { it.categoryId }
                .mapValues { (_, subs) -> subs.sumOf { it.toMonthlyAmount() } }

            val total = breakdown.values.sum().coerceAtLeast(1.0)

            breakdown.entries
                .mapNotNull { (categoryId, amount) ->
                    val category = categoryMap[categoryId] ?: return@mapNotNull null
                    CategorySpend(
                        category = category,
                        monthlyAmount = amount,
                        percentage = (amount / total).toFloat()
                    )
                }
                .sortedByDescending { it.monthlyAmount }
        }

    private fun Subscription.toMonthlyAmount(): Double =
        when (frequency) {
            BillingFrequency.WEEKLY -> amount * 4
            BillingFrequency.MONTHLY -> amount
            BillingFrequency.QUARTERLY -> amount / 3
            BillingFrequency.SEMI_ANNUAL -> amount / 6
            BillingFrequency.ANNUAL -> amount / 12
            BillingFrequency.CUSTOM -> amount
        }
}
